Feature/checkout

lesser1 opened the pull request:
This is a relatively big pull request. I'm sorry about that.

First of all, [this is the Gropius issue](https://frontend.gropius.duckdns.org/components/9d12d6ac-e85a-407c-bdff-a8321fb75d3a/issues/0d686183-befe-48a6-92c3-549392d4f9bc).

Please consider that
- The user can cancel the checkout by clicking the CANCEL button or by clicking one of the buttons in the topmost app bar. The CANCEL button has a confirmation dialog -- the user has to confirm that they actually want to cancel the checkout. The buttons in the app bar do not have that. I would suggest I open an extra issue for that in order to keep this feature implementation as concise as possible.
- In the backend, the prices of the products are in Cents not in Euros. Meanwhile, as far as I know, the shipping fees are in Euros. The frontend still does not convert the product prices meaning it does not simply divide them by 100. I want to do that in a separate issue for the whole application. This means that the costs displayed during checkout are of course not correct.
- The project supervisors requested that as a user, one can provide the VAT identification number during checkout. I have yet to implement that. I would suggest that this should be part of an extra issue. Also the backend has to be extended to take care of that information.
- When testing the checkout, you need to have added a shipment method. Right now, there is no way to do that via the Frontend. Instead, you have to execute a GraphQL mutation via GraphiQL:
```
mutation MyMutation {
  createShipmentMethod(
    input: {baseFees: 10, description: "DHL kinda rules the world.", externalReference: "", feesPerItem: 10, feesPerKg: 10, name: "DHL"}
  ) {
    id
  }
}
```
Don't forget to set the authorization header:
```
{
  "authorization": "Bearer [paste your token here]"
}
```

The following issues regarding the feature implementation have already been found / tracked:
- [x] The SelectOrAddNewAddressCards and SelectOrAddNewCreditCardCard should not be as wide.
- [ ] The shipment methods are not cached, but queried anew each time when switching from address to shipment. -- into a separate issue
- [ ] Clicking on the product in the ShipmentInformationView navigates directly to the product without a "Confirm cancellation" prompt. -- into a separate issue
- [x] Explanations of what Invoice and Prepayment are would be helpful.
- [x] While the order is being created, there should be an animation indicating to the user that something is happening in the background.

### Definition of Done
- [x] Requirements of the issue are met
- [x] Code has been reviewed
- [x] Code is documented
- [x] GraphQL related artefacts are documented

lesser1 committed:
Introduce routeNames.ts

- Add routeNames.ts storing the route names
- Rename routes according to Vue Router standards
- Define route names for new routes related to checkout
- Replace hard coded route name strings throughout the project
  with routeNames references

lesser1 committed:
Add model, Store state of order information

- Add interfaces and classes regarding checkout / order information
- Add orderInformation state ref to the app store

lesser1 committed:
Add new route name

lesser1 committed:
Merge branch 'maintenance/new-inventory-status-query' into feature/checkout

lesser1 committed:
Lint and format

lesser1 committed:
Merge branch 'maintenance/new-inventory-status-query' into feature/checkout

lesser1 committed:
Merge branch 'main' into feature/checkout

lesser1 committed:
Add checkout stage for address information

- Add new routes for the checkout stages
- Add checkout layout: TheCheckoutLayout.vue
- Add new component for selecting or saving a (new) address:
  SelectOrAddNewAddressCard.vue
- Note that the checkout stages payment information and summary are
  yet to be implemented
- Add error messages
- Add new rule for when a string input is required
- Add new access right for checkout process: canAccessCheckout
- Rearrange access rights alphabetically
- Modify ShoppingCartView.vue to allow user to enter checkout phase by
  clicking "PROCEED TO CHECKOUT" button
- Add graphql query and mutation for querying active addresses of the
  current user and adding a new address

lesser1 committed:
Add initial model for payment information

lesser1 committed:
Add payment info complete check

- Ensure billing address of order info is set if delivery address is set
- Add check whether or not user can proceed after adding payment info

lesser1 committed:
Merge branch 'main' into feature/checkout

lesser1 committed:
Add representation of shipment method to the model

- Add ShipmentMethod interface
- Add ShipmentMethodImpl class
- Extend OrderInformation to include ShipmentMethod
- Change order or properties of OrderInformation to represent the
  information flow of checkout process
- Extend method in app.ts -- resetOrderInformationToUndefined() --
  to also reset the shipment method

lesser1 committed:
Add checkout steps shipment info and payment info

lesser1 committed:
Refine model

- Add interface and implementation OrderItem (OrderItemImpl)
- Rename OrderInformation to just Order
- Remove shipment method from Order
- Apply renaming of Order(Information) to various code files
- Simplify resetOrderInformationToUndefined action in app.ts
  - rename to resetOrder
  - just reset to order to empty OrderImpl object

lesser1 committed:
Add view for shipment information at checkout

- Add ID of product variant version to ShoppingCartItem because it is
  needed for calculating the shipment fees
- Add view that lets user choose a shipment method for each order item

lesser1 committed:
Add request to create order

- Add GraphQL mutation createOrder and corresponding error message
- Add prop for order ID to OrderSummaryView.vue
- Pass ID of created order to order summary view

lesser1 committed:
Add initial order summary

- Add OrderCard.vue SFC: This card displays some information regarding
  a specific order for a specific purpose, e.g., Summary
- Add OrderSummaryView.vue that simply shows one OrderCard for the
  summary of the order that has just been placed.
- Add orderStatus enum for the order status that comes from the order
  service. Also add mapping code for mapping from graphql code-
  generated enum to new frontend enum.
- Add graphql query for getting an order by ID.

lesser1 committed:
Merge branch 'main' into feature/checkout

lesser1 committed:
Improvements

- Add dialog that lets the user confirm or cancel something:
  ConfirmOrCancelDialog.vue.
- Display total amount to pay to the user when the user has to select
  payment information.
- Add methods that calculate the total cost of the order based on the
  shopping cart items and the shipping fees.
- User has to confirm both cancellation and making of order via a
  dialog.

lesser1 committed:
Add success notification for adding credit card

- Also, add missing import in SelectOrAddNewAddressCard.vue.

lesser1 committed:
Styling: Set width of checkout stages

lesser1 committed:
Add progress circle during order creation

- Add a persistent overlay to the screen when the app is awaiting
  the order service to respond to the order creation request.
- The overlay simply displays an indeterminate prgress circle that
  shows a circular loading animation.

lesser1 committed:
Add descriptions to payment methods

legendofa commented on src/layouts/checkout/TheCheckoutLayout.vue:
```diff
@@ -0,0 +1,346 @@
+    }
+    isAwaitingOrderCreation.value = true
+    try {
+        var idOfCreatedOrder = await createOrder(
```
The order also needs to be placed via the GraphQL endpoint here, the idea was to create the order before showing the summary and placing it after the user has used the confirm button on the summary page

lesser1 replied:
Ah, okay, that explains a lot! I will change it.

legendofa reviewed (changes requested):
I tested the process of creating an order and it looks fine/works as i expected. Just one thing is missing: Invoking the endpoint to place the order

lesser1 committed:
Modify rule that checks expiration date string

- Remove text-capitalize class from the fields Number and Expiration
  Date because it would result in weird capitalization of validation
  error hints.
- Extend rule isValidCreditCardExpirationDate to check whether the date
  is a future date.

lesser1 committed:
Add parsing of expiration date field input

- MM/YYYY gets parsed to MM/YY

lesser1 committed:
Add order placement

- Rename CreateOrder mutation to createOrder
- Extend functionality of OrderCard
- Add extra step of placing order

legendofa reviewed (approved):
The order placement logic works now.

However, it would be nice to see an overview of the order before pressing the placement button. I can understand if this is not possible due to time constraints, so i will approve it for now. I think the user would expect a redirect to a simple orders overview page, which shows all orders in a list. And (maybe) the same order overview page component as before placing the order + the deliver status. 

Overall i like the look and feel how the order process is now implemented. Good job!

lesser1 commented:
> The order placement logic works now.
> 
> However, it would be nice to see an overview of the order before pressing the placement button. I can understand if this is not possible due to time constraints, so i will approve it for now. I think the user would expect a redirect to a simple orders overview page, which shows all orders in a list. And (maybe) the same order overview page component as before placing the order + the deliver status.
> 
> Overall i like the look and feel how the order process is now implemented. Good job!

Thank You! Regarding the overview, you are absolutely right: It would be really nice but I would argue there is not enough time right now. I will resolve any conflicts regarding the branch and merge it into main afterwards.

lesser1 committed:
Merge branch 'main' into feature/checkout
